audeer/core/utils.py

```diff
@@ -61,6 +61,62 @@ def new_func(*args, **kwargs):
     return _deprecated
 
 
+def deprecated_keyword_argument(
+        *,
+        deprecated_argument: str,
+        removal_version: str,
+        alternative_argument: str = None,
+) -> Callable:
+    r"""Mark keyword argument as deprecated.
+
+    Provide a `decorator <https://www.python.org/dev/peps/pep-0318/>`_
+    to mark keyword arguments as deprecated.
+
+    You have to specify the version,
+    for which the deprecated argument will be removed.
+    The content assigned to ``deprecated_argument``
+    is passed on to the ``alternative_argument``.
+
+    Args:
+        deprecated_argument: keyword argument to be marked as deprecated
+        removal_version: version the code will be removed
+        alternative_argument: keyword argument that should be used instead
+
+    Example:
+        >>> @deprecated_keyword_argument(
+        ...     deprecated_argument='foo',
+        ...     alternative_argument='bar',
+        ...     removal_version='2.0.0',
+        ... )
+        ... def function_with_new_argument(*, bar):
+        ...     pass
+
+    """
+    def _deprecated(func):
+        # functools.wraps preserves the name
+        # and docstring of the decorated code:
+        # https://docs.python.org/3/library/functools.html#functools.wraps
+        @functools.wraps(func)
+        def new_func(*args, **kwargs):
+            if deprecated_argument in kwargs:
+                message = (
+                    f"'{deprecated_argument}' argument is deprecated "
+                    f"and will be removed with version {removal_version}."
+                )
+                if alternative_argument is not None:
+                    message += f" Use '{alternative_argument}' instead."
+                    argument_content = kwargs.pop(deprecated_argument)
```

Comment:
Is it a good idea to only `pop` if an alternative argument is given? If this is not the case, it will probably raise an error.

Reply:
Another problem may arise if there is no direct mapping between the old and the new argument. Maybe we should also give the option to pass a function that would do the mapping.

Reply:
> Is it a good idea to only `pop` if an alternative argument is given? If none is provided, it will probably raise an error later.
This way we can support removal of keyword arguments. If `alternative_argument=None` this is exactly this case.

Reply:
But what if I want I only want to remove the keyword but not assign to an alternate argument?

Reply:
This is exactly what I'm talking about.

Reply:
Sorry, I'm lost. But in that case the keyword is not popped.

Reply:
```python
@audeer.deprecated_keyword_argument(
    deprecated_argument='foo',
    alternative_argument='bar',
    removal_version='1.0.0',
) 
```
will provide an alternative keyword argument `bar` instead of `foo`.
```python
@audeer.deprecated_keyword_argument(
    deprecated_argument='foo',
    removal_version='1.0.0',
) 
```
will just remove the kyword argument `foo`

Reply:
I.e. the keyword is still passed on to the function, which will probably raise an error.

Reply:
Ah, ok, now I get your point. We still need to `pop` ;)

Reply:
Yes, that's what I meant when I asked if it is a good idea to only pop if an alternate argument is given :)

Reply:
I moved the `pop` line now before the `if`-statement.

Reply:
As far as I understand this decorater works  not like `deprecated` where you still have the function, but here we assume that the deprecated argument is actually removed from the parameters.

Reply:
> I moved the pop line now before the if-statement.

Thanks. What about an optional function that maps the old to the new argument?

Reply:
I renamed `alternative_argument` to `new_argument` and added a mapping argument, to map between the deprecated and the new argument.